ASSERTION FAILED: scrollerImp == scroller->scrollerImp() if closing a window while scrolling

nmoucht opened the pull request:
#### f719f87b07476ed574a5a450bf39931218d52798
<pre>
ASSERTION FAILED: scrollerImp == scroller-&gt;scrollerImp() if closing a window while scrolling
<a href="https://bugs.webkit.org/show_bug.cgi?id=255886">https://bugs.webkit.org/show_bug.cgi?id=255886</a>
rdar://108469342

Reviewed by Simon Fraser.

The order of destruction is ScrollingTreeScrollingNodeDelegateMac -&gt; ScrollerPairMac -&gt;
ScrollerMac. We don&apos;t want to release the scroller imps in ScrollingTreeScrollingNodeDelegateMac
because we have yet to remove the delegate from the scroller imps. This can lead to a state where
we get a call to convertContentPoint when our reference to the scroller imps in ScrollerMac are
already released. To resolve this, move this call to ~ScrollerPairMac and move the ScrollerMac
calls to the individual scroller imps to their own function and call them in ~ScrollerPairMac.

* Source/WebCore/page/scrolling/mac/ScrollerPairMac.mm:
(-[WebScrollerImpPairDelegateMac scrollerImpPair:convertContentPoint:toScrollerImp:]):
(WebCore::ScrollerPairMac::~ScrollerPairMac):
* Source/WebCore/page/scrolling/mac/ScrollingTreeScrollingNodeDelegateMac.mm:
(WebCore::ScrollingTreeScrollingNodeDelegateMac::~ScrollingTreeScrollingNodeDelegateMac):

Canonical link: <a href="https://commits.webkit.org/263557@main">https://commits.webkit.org/263557@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/833b28758441bc25d751058fd85405a49b74e573

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/4418 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/4538 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/4670 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/5897 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/4619 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/4407 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/4706 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/4504 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/4846 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/4483 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/4647 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/3971 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/5930 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/2142 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/3948 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/8319 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/3987 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/4017 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/5588 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/4427 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/3603 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/3950 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/3947 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/8002 "Built successfully") | | | 
| [❌ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/567 "Failed to push commit to Webkit repository") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/4305 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

whsieh commented on Source/WebCore/page/scrolling/mac/ScrollerPairMac.mm:
```diff
@@ -99,6 +99,7 @@ - (NSPoint)scrollerImpPair:(NSScrollerImpPair *)scrollerImpPair convertContentPo
         scroller = &_scrollerPair->horizontalScroller();
     else
         scroller = &_scrollerPair->verticalScroller();
+    WTFLogAlways("convertContentPoint node:%d", _scrollerPair->node().scrollingNodeID());
```
Nit - some stray debug logging.

nmoucht replied:
Fixed.

whsieh commented on Source/WebCore/page/scrolling/mac/ScrollerPairMac.mm:
```diff
@@ -145,11 +146,14 @@ - (void)scrollerImpPair:(NSScrollerImpPair *)scrollerImpPair updateScrollerStyle
 
 ScrollerPairMac::~ScrollerPairMac()
 {
+    WTFLogAlways("ScrollerPairMac::~ScrollerPairMac %d", node().scrollingNodeID());
```
(Ditto)

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/page/scrolling/mac/ScrollerPairMac.mm:
```diff
@@ -147,9 +147,14 @@ - (void)scrollerImpPair:(NSScrollerImpPair *)scrollerImpPair updateScrollerStyle
 {
```
Should we assert that we're on the main thread here?

nmoucht replied:
Right, I think we should.

nmoucht replied:
Actually I'm seeing this assert fail in the tests. This destruction is coming from the ScrollingTree being destructed. Does it make sense for this to happen off the main thread? Regardless we have the explicit bounce to main thread call here.

smfr commented on Source/WebCore/page/scrolling/mac/ScrollerPairMac.mm:
```diff
@@ -147,9 +147,14 @@ - (void)scrollerImpPair:(NSScrollerImpPair *)scrollerImpPair updateScrollerStyle
 
     ensureOnMainThread([scrollerImpPair = std::exchange(m_scrollerImpPair, nil)] {
     });
+    releaseReferencesToScrollerImpsOnTheMainThread();
```
Seems weird to have two separate bounces to the main thread here. Can we combine them?

nmoucht replied:
Fixed.

nmoucht committed:
ASSERTION FAILED: scrollerImp == scroller->scrollerImp() if closing a window while scrolling
https://bugs.webkit.org/show_bug.cgi?id=255886
rdar://108469342

Reviewed by Simon Fraser.

The order of destruction is ScrollingTreeScrollingNodeDelegateMac -> ScrollerPairMac ->
ScrollerMac. We don't want to release the scroller imps in ScrollingTreeScrollingNodeDelegateMac
because we have yet to remove the delegate from the scroller imps. This can lead to a state where
we get a call to convertContentPoint when our reference to the scroller imps in ScrollerMac are
already released. To resolve this, move this call to ~ScrollerPairMac and move the ScrollerMac
calls to the individual scroller imps to their own function and call them in ~ScrollerPairMac.

* Source/WebCore/page/scrolling/mac/ScrollerPairMac.mm:
(-[WebScrollerImpPairDelegateMac scrollerImpPair:convertContentPoint:toScrollerImp:]):
(WebCore::ScrollerPairMac::~ScrollerPairMac):
* Source/WebCore/page/scrolling/mac/ScrollingTreeScrollingNodeDelegateMac.mm:
(WebCore::ScrollingTreeScrollingNodeDelegateMac::~ScrollingTreeScrollingNodeDelegateMac):

Canonical link: https://commits.webkit.org/263557@main
